src/Mailchimp/Manager.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+namespace AppBundle\Mailchimp;
+
+use AppBundle\Entity\Adherent;
+use AppBundle\Entity\AdherentMessage\AdherentMessageInterface;
+use AppBundle\Mailchimp\Campaign\CampaignRequestBuilder;
+use AppBundle\Mailchimp\Exception\InvalidCampaignIdException;
+use AppBundle\Mailchimp\Synchronisation\Command\AdherentCommandInterface;
+use AppBundle\Mailchimp\Synchronisation\RequestBuilder;
+use Psr\Container\ContainerInterface;
+use Psr\Log\LoggerAwareInterface;
+use Psr\Log\LoggerAwareTrait;
+
+class Manager implements LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    private $driver;
+    private $requestBuildersLocator;
+
+    public function __construct(Driver $driver, ContainerInterface $requestBuildersLocator)
+    {
+        $this->driver = $driver;
+        $this->requestBuildersLocator = $requestBuildersLocator;
+    }
+
+    /**
+     * Creates/updates a Mailchimp member
+     */
+    public function editMember(Adherent $adherent, AdherentCommandInterface $message): void
+    {
+        $requestBuilder = $this->requestBuildersLocator
+            ->get(RequestBuilder::class)
+            ->updateFromAdherent($adherent)
+        ;
+
+        $result = $this->driver->editMember(
+            $requestBuilder->buildMemberRequest($message->getEmailAddress())
+        );
+
+        if ($result) {
+            $this->logger->info(sprintf('Mailchimp member "%s" has been edited', $adherent->getUuidAsString()));
+
+            // Active/Inactive member's tags
+            $result = $this->driver->updateMemberTags(
+                $requestBuilder->createMemberTagsRequest($message->getEmailAddress(), $message->getRemovedTags())
+            );
+
+            if ($result) {
+                $this->logger->info(sprintf('Mailchimp member "%s" tags have been updated', $adherent->getUuidAsString()));
+            }
+        }
+    }
+
+    public function getCampaignContent(AdherentMessageInterface $message): string
+    {
+        if (!$message->getExternalId()) {
+            throw new InvalidCampaignIdException(sprintf('Message "%s" does not have a valid campaign id', $message->getUuid()));
+        }
+
+        return $this->driver->getCampaignContent($message->getExternalId());
+    }
+
+    public function editCampaign(AdherentMessageInterface $message): bool
+    {
+        $requestBuilder = $this->requestBuildersLocator->get(CampaignRequestBuilder::class);
+
+        $editCampaignRequest = $requestBuilder->createEditCampaignRequestFromMessage($message);
+
+        if (!$campaignId = $message->getExternalId()) {
+            if (!$campaignId = $this->driver->createCampaign($editCampaignRequest)) {
+                throw new \RuntimeException(sprintf('Campaign for the message "%s" was not created', $message->getUuid()));
```

Comment:
`Campaign for the message "%s" has not been created`

Reply:
done